src/Control/Distributed/Process.hs

```diff
@@ -402,12 +410,27 @@ spawnChannelLocal proc = do
 callLocal ::
      Process a  -- ^ Process to run
   -> Process a  -- ^ Value returned
-callLocal proc = do
-  mv <- liftIO newEmptyMVar
-  self <- getSelfPid
-  _ <- spawnLocal $ do
-     link self
-     try proc >>= liftIO . putMVar mv
-  liftIO $ takeMVar mv >>=
-     either (throwIO :: SomeException -> IO a) return
+callLocal proc = mask $ \release -> do
+  lock  <- liftIO $ newMVar ()
+  tidMV <- liftIO $ newEmptyMVar
+  mv    <- liftIO newEmptyMVar
+  _ <- spawnLocal $ mask $ \release' -> do
+         liftIO $ myThreadId >>= putMVar tidMV
+         ep <- try $ release' $ proc <* liftIO (takeMVar lock)
+         liftIO $ putMVar mv ep
+  tid <- liftIO $ takeMVar tidMV
+  let fetchResult =
+        (takeMVar mv >>= either (throwIO :: SomeException -> IO a) return)
+        `Exception.catch`
+            (\e -> Exception.mask $ \release' -> do
+                     ml <- tryTakeMVar lock
+                     case ml of
+                        Nothing -> -- lock already taked by worker so we need
+                                   -- to rethrow exception now
+                          throwIO e
+                        Just{}  -> do
+                          throwTo tid (e::SomeException)
+                          putMVar lock ()
+                     release' fetchResult)
+  release $ liftIO $ fetchResult
```

Comment:
Since `fetchResult` is interruptible there is no need to unmask exceptions. Then the mask in the `catch` handler becomes unnecessary too.

`throwTo` is an interruptible operation, so should be protected with an uninterruptible mask or the spawned thread could leak.

`unmask` seems a better name than `release`.